File: backend/app/core/upgrades/stat_modifiers/bullet_size_2.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class BulletSize2(UpgradeImplementation):
    id = 'bullet_size_2'
    name = 'Heavy Rounds'
    description = '+25% bullet size'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.25
        appliers.modifiers.add_multiplier('bullet', 'size', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True

File: frontend/src/game/upgrades/stat_modifiers/bullet_size_2.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const BulletSize2Def: UpgradeDef = {
  id: "bullet_size_2",
  name: "Heavy Rounds",
  description: "+25% bullet size",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Bullet,
  fieldInTargetClass: UpgradeStatID.Size,
  value: 0.25,
  isMultiplier: true,
  specificAttackType: "bullet",
  stackable: true,
  maxStacks: 2,
}

export class BulletSize2 extends Upgrade {}
